source/portable/NetworkInterface/RM46L852/NetworkInterface.c

```diff
@@ -0,0 +1,369 @@
+                                        configMINIMAL_STACK_SIZE,
+                                        NULL,
+                                        configMAX_PRIORITIES - 1,
+                                        &receiveTaskHandle );
+
+            if( ( receiveTaskHandle == NULL ) || ( xTaskCreated != pdPASS ) )
+            {
+                FreeRTOS_printf( ( "Failed to create the handler task." ) );
+
+            }
+            /* After this, the task should not be created. */
+            else
+                xFirstCall = pdFALSE;
+        }
+
+        return pdPASS;
+    }
+    else
+    {
+        return pdFAIL;
+    }
+}
+
+/*
+ *  This function is used to call EMAC API to transmit packets.
+ *  And freeing the network buffers.
+ */
+
+BaseType_t xNetworkInterfaceOutput( NetworkBufferDescriptor_t * const pxNetworkBuffer,
+                                    BaseType_t xReleaseAfterSend )
+{
+      struct pbuf_struct xpbuf;
+      xpbuf.payload = pxNetworkBuffer->pucEthernetBuffer;
+      xpbuf.len = pxNetworkBuffer->xDataLength;
+      xpbuf.tot_len = pxNetworkBuffer->xDataLength;
+      xpbuf.next = NULL;
+      if(xpbuf.tot_len < MIN_PKT_LEN)
+      {
+          xpbuf.tot_len = MIN_PKT_LEN;
+          xpbuf.len = MIN_PKT_LEN;
+      }
+
+      taskENTER_CRITICAL();
+      {
+          EMACTransmit(&hdkif_data[0U],&xpbuf);
+      }
+      taskEXIT_CRITICAL();
+
+      if( xReleaseAfterSend == pdTRUE )
+      {
+          vReleaseNetworkBufferAndDescriptor( pxNetworkBuffer );
+      }
+      return pdPASS;
+}
+
+/*
+ *  This Task is notified everytime a packet is received.
+ *  It calls EMACReceive for actual packet processing.
+ */
+
+void prvEMACHandlerTask( void * pvParameters  )
+{
+
+    while( pdTRUE )
+    {
+        ulTaskNotifyTake( pdTRUE, pdMS_TO_TICKS( 500 ) );
+
+        taskENTER_CRITICAL();
+        {
+            EMACReceive(&hdkif_data[0U]);
+        }
+        taskEXIT_CRITICAL();
+    }
+}
+
+/*
+ *  This function sends RX event to IP task for further processing the data packet.
+ */
+
+void prvProcessFrame( NetworkBufferDescriptor_t * pxBufferDescriptor)
+{
+
+    if( pxBufferDescriptor != NULL )
+    {
+
+        if( ipCONSIDER_FRAME_FOR_PROCESSING( pxBufferDescriptor->pucEthernetBuffer ) == eProcessBuffer )
+        {
+
+            /* send packet to IP task for further processing */
+            IPStackEvent_t xRxEvent;
+            xRxEvent.eEventType = eNetworkRxEvent;
+            xRxEvent.pvData = ( void * ) pxBufferDescriptor;
+            if( xSendEventStructToIPTask( &xRxEvent, 0 ) == pdFALSE )
+            {
+                vReleaseNetworkBufferAndDescriptor( pxBufferDescriptor );
+                iptraceETHERNET_RX_EVENT_LOST();
+                FreeRTOS_printf( ( "RX Event Lost\n" ) );
+            }
+        }
+        else
+        {
+            #if ( ( ipconfigHAS_DEBUG_PRINTF == 1 ) && defined( FreeRTOS_debug_printf ) )
+                const EthernetHeader_t * pxEthernetHeader;
```

Comment:
It's against C coding conventions to declare variables in the middle of the function. These variables should be wrapped in an ifdef at the start of the function.

Reply:
Hello @Skptak, as the variable ` pxEthernetHeader` is only used for logging, I prefer to see a local declaration.

Normally we ( in FreeRTOS+TCP ) adhere to the rule that local declarations are OK, unless they're mixed with code, They should be placed at the beginning of a compound block:

~~~
    if ()
    {
        unsigned rc;  // good
        counter++;
    }
    if ()
    {
        counter++;
        unsigned rc;  // bad: code followed by a declaration
    }
~~~

The advantage is that the code will be easier to read. The reader doesn't have don't have to scroll back to see the declarations. Another advantage is that the scope of the variable is limited when declared in a local block.